src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -326,108 +310,74 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
-                    }
-                    analyzedArguments.Free();
-                    indexerGroup.Free();
-                    goto done;
+            receiverPlaceholder = new BoundListPatternReceiverPlaceholder(node, GetValEscape(inputType, inputValEscape), inputType) { WasCompilerGenerated = true };
+            bool hasErrors = false;
+            if (inputType.IsSZArray())
+            {
+                hasErrors |= !TryGetSpecialTypeMember(Compilation, SpecialMember.System_Array__Length, node, bindingDiagnostics, out PropertySymbol lengthProperty);
+                if (lengthProperty is not null)
+                {
+                    lengthAccess = new BoundPropertyAccess(node, receiverPlaceholder, lengthProperty, LookupResultKind.Viable, lengthProperty.Type);
+                }
+                else
+                {
+                    lengthAccess = new BoundBadExpression(node, LookupResultKind.Empty, ImmutableArray<Symbol?>.Empty, ImmutableArray<BoundExpression>.Empty, CreateErrorType(), hasErrors: true);
                 }
-                lookupResult.Clear();
+            }
+            else
+            {
+                hasErrors |= !TryBindLengthOrCount(node, receiverPlaceholder, receiverPlaceholder.Type, out lengthAccess, bindingDiagnostics);
             }
 
-            // If the argType is missing or the indexer lookup has failed, we will fallback to the implicit indexer support.
-            found = TryLookupLengthOrCount(receiverType, lookupResult, out lengthProperty, ref useSiteInfo) &&
-                    TryFindIndexOrRangeIndexerPattern(lookupResult, receiverOpt: null, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo);
+            if (lengthAccess is null)
+            {
+                Error(bindingDiagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+            }
+            else
+            {
+                CheckValue(lengthAccess, BindValueKind.RValue, diagnostics);
+            }
+
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var systemIndexType = GetWellKnownType(WellKnownType.System_Index, bindingDiagnostics, node);
+            argumentPlaceholder = new BoundListPatternUnloweredIndexPlaceholder(node, systemIndexType) { WasCompilerGenerated = true };
+            analyzedArguments.Arguments.Add(argumentPlaceholder);
 
-done:
-            if (found)
+            indexerAccess = BindElementAccessCore(node, receiverPlaceholder, analyzedArguments, bindingDiagnostics);
+            indexerAccess = CheckValue(indexerAccess, BindValueKind.RValue, bindingDiagnostics);
+            Debug.Assert(indexerAccess is BoundIndexerAccess or BoundIndexOrRangePatternIndexerAccess or BoundArrayAccess or BoundBadExpression or BoundDynamicIndexerAccess);
+            analyzedArguments.Free();
+
+            if (bindingDiagnostics.AccumulatesDiagnostics && bindingDiagnostics.HasAnyErrors())
```

Comment:
>HasAnyErrors [](http://example.com/codeflow?start=80&length=12)

Why do we care about this at all? At the very least, I think we should be using ```HasAnyResolvedErrors``` instead. I think we shouldn't have this logic at all. Result of this function shouldn't depend on whether we collect diagnostics or not. Obsolete errors also shouldn't have any effect.